### PR TITLE
New Hook System

### PR DESCRIPTION
**Pull request by FireSiku**
Alright, submitting PR For the big change that has had so much discussion about.

The system is backward-compatible with the current one, but some more testing will probably be required today/tomorrow. I'm opening this PR because I feel like its really at a state where I feel good about it.

Not looking for an immediate merge, but since its mostly finalized aside from some details, i'm putting this here for code review ;)

When it'll finally be ready to merge, I'll rename the file to hooks.lua, but i wanted to keep them separate to prevent git showing a diff of the old and new systems.

### CODE REVIEW

File: vmf/scripts/mods/vmf/modules/core/newhooks.lua [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    if type(method) ~= "string" then
+        self:error("(hook): 'method' - string expected, got %s", type(method))
+        return
+    end
+
+    if obj then
+        -- obj can be a string. We'll need to grab the actual object first.
+        -- if we can't find object, we don't need to go any further.
+        if type(obj) == "string" then
+            if not _G[obj] then return end
```
Comment (unshame) [inline vmf/scripts/mods/vmf/modules/core/newhooks.lua]: Silent fail?

**Comment by unshame** [comment]
I might've missed something so disregard if this has been discussed already and an agreement was made but I thought we'd agreed that this implementation of before and after hooks isn't gonna be a thing? Also without the ability to set multiple hooks to the same function (which I think bi correctly said is a bad idea), hook objects are completely useless. Also applying a new hook should replace the old one, not throw an error.

**Comment by FireSiku** [comment]
Honestly, this is one of the reasons I wanted to make the PR and probably shouldve made it earlier for the sake of discussions. On your points:



On the subject of different hook types (before / after, etc):
I'm still strongly in support of this as a way to reduce the dependency on load order that we have. Considering we have to deal with people that can't even have the common sense to put VMF at the top, the more tools we have to allow any mod to be loaded in any order, the better we are. 

There was a lot of discussion on before/prehooks being _always_ called before the hook chain, bi managed to convince me that it wouldn't actually solve anything toward the goal of alleviating ordering issues, so those are now part of the hook chain like any regular hooks, and their benefits is simply that they automatically call the original function without having to worry about breaking stuff (for example if you simply call `func(...)`, but the function is supposed to return a value and you end up returning nil instead.)

After-hooks on the other hand have proven that a syntactic version would be incredibly inefficient due to having to handle return values, and a non-syntactic version has lots of performance applilcations, without having the "abuse" problem of before-hooks. For example, take three mods with hooks like this, called A, B, C:
```lua
function(func, self, value, ...)
    func(self, value, ...)

    mod.update_internals(self)
end
```
The stack would execute C, which calls B, then A, then the original, then back to A, back to B, back to C. (Before-hooks make use of tail calls to avoid this). Having them executed afterward the hook chain has resolved makes it rather easy linear execution, we can even skip a few steps!
```lua
-- Instead of this
mod:after(Object, "Method", function(self, value, ...)
    mod.update_internals(self)
end)
-- We can do this
mod:after(Object, "Method", mod.update_internals)
```

TL;DR: Before implementation was question, rightfully so and has been integrated in the hook chain. After implementation was never a problem and even improves things.



On the subject of hook objects, it turns out they still have some use! I personally even disconnected a bit from the idea as I worked on it, but I never removed it because it was never in the way. When I was working on it, I never felt like I had to take hook objects into considerations, their code is cleanly encapsulated, they have a very small footprint and should the modder completely ignore them, immediately get garbage collected (they're not even kept around internally)

Then I decided to do a test on a modder deciding to use hook object to enable/disable a hook as needed, vs calls to `mod:hook_enable`. Here are the results:
```
100,000 iterations
hook_obj:enable - 2ms
vmf:hook_enable("PingTargetExtension.set_pinged") - 171ms
vmf:hook_enable(PingTargetExtension, "set_pinged") - 111ms
```
So, it seems like they do have _some_ uses, and they do not hinder maintainability, which I'm all for keeping at that point.




Last two points (holy shit I'm still going)
1. The current implement allows you to have both a hook as part of the chain (before/hook) and a hook after the chain (after) on the same function. I'm going to be removing that later tonight.

2. Yeah, calling a hook with the same function throws an error that you cant hook a function twice instead of overwriting it... but I couldn't find anyone that used that functionality in the first place (unless I missed it), so the only ones that would be doing it in the future could be fresh blooded modders who thinks you can hook the same function multiple times we arent stopping them (I know modders are supposed to be smarter than end users, but you never know). 
I'm skeptical that the functionality is ever _required_ for an advanced modder's desired effect. Feel free to link me to a practical example.

**Comment by unshame** [comment]
I just realized I wasn't looking at the latest commit.  

#### Hook objects
> On the subject of hook objects, it turns out they still have some use!  

They don't, they just work slightly faster. That's not a usecase. And performance of enabling/disabling hooks isn't a concern.  
Since the one-hook-per-function-per-mod limitation stays, hook objects should go. They're completely redundant.

#### Before/after/raw hooks

```
function VMFMod:hook_enable(obj, method)    generic_hook_toggle(self, obj, method, "hook_enable") end
function VMFMod:before_enable(obj, method)  generic_hook_toggle(self, obj, method, "before_enable") end
function VMFMod:after_enable(obj, method)   generic_hook_toggle(self, obj, method, "after_enable") end
function VMFMod:rawhook_enable(obj, method) generic_hook_toggle(self, obj, method, "rawhook_enable") end

function VMFMod:hook_disable(obj, method)    generic_hook_toggle(self, obj, method, "hook_disable") end
function VMFMod:before_disable(obj, method)  generic_hook_toggle(self, obj, method, "before_disable") end
function VMFMod:after_disable(obj, method)   generic_hook_toggle(self, obj, method, "after_disable") end
function VMFMod:rawhook_disable(obj, method) generic_hook_toggle(self, obj, method, "rawhook_disable") end
```
What? We have one hook per function, it should just be hook_disable/hook_enable and that's it.
Other functions should be called
hook
hook_after
hook_before
hook_raw (hook_origin makes more sense imo)    
for consistency's sake.  

#### After hook
The fact that after hooks are the only ones able to ignore broken hook chains isn't a good thing, it's just another inconsistency. What sense does it make that an after hook works completely different than a before hook? Since there are rawhooks now, broken hook chains should be fixed, not worked around.   
And again, their improved performance isn't a usecase. They should either be a part of the hook chain and depend on the load order like the rest, or be removed if the performance is too bad.  
Also, I don't know if reversing the order in which after hooks execute is a good idea.

#### Hooking same function multiple times
I'm okay if it throws an error, that's probably a better way of doing it.  Just pointing out that this is not backwards compatible (but if no one did it, then it doesn't matter).

**Comment by Azumgi** [comment]
I decided to think a little bit about it and came up with the solution for hook_after that shouldn't affect performance much.
```lua
local function return_vararg(...)
  return ...
end

local next_hook -- next hook function (can also be an original function)
local current_after_hook -- a hook function which is passed in 'mod:hook_after'

local hook_function = function(...)
  return return_vararg(next_hook(...), current_after_hook(...))
end
```
Cons:
- If a modder will set `return` in their `after_hook_function`, it will be applied to other return values of an original function. But it doesn't really matter since in the end they won't be used anyway.

Pros:
- It will allow put all hooks in a single chain.
- Hooking system can be rewritten to no use `fn` function (which is created inside `create_internal_hook` function), thus avoiding 1 unnecessary line in error traceback. But I don't know if it's possible in your implementation.

**Comment by IamLupo** [comment]
Yeah, when i made hooking system version 2 for V1 i made it compatable for multiple hooking. I never implemented before and after because sometimes you want to not execute the original function. 

A question: In your functionality you say you must have a obj and method. Like you expect that the function has to be in a object. When the obj is nil it still works?

You also have a overview what mods have hooked what? Like there is a debug parameter to keep track of this.

**Comment by FireSiku** [comment]
This is the function signature: `:hook([object], method, handler)`
which means it supports these methods of calling a hook:
The system currently support these methods of calling a hook:
```lua
mod:hook("CrosshairUI.draw", hook_function) -- VT1 style
mod:hook("UIResolution", hook_function) -- Hooking a global function
mod:hook("CrosshairUI", "draw", hook_function) -- Needed for delayed hooks.
mod:hook(CrosshairUI, "draw", hook_function) -- Hooking with object
```
obj being nil (when obj and method are correctly split and detected) is taken into account.

We can easily have a "debug map" of who hooked what by iterating over the internal _registry table.

**Comment by FireSiku** [comment]
So, I've done all the changes discussed on discord. Got rid of before/after, kept hook_origin and hook_safe.

It was to be ready for merge, but i found some last minute issues that I'm correcting.

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 2]
```diff
@@ -1,367 +1,403 @@
...
+            method, obj = split_function_string(obj)
+        else
+            self:error("(%s): trying to toggle hook without giving a method name. %s", func_name)
+        end
+    end
 
-    create_hook_entry(hook_info[1], hooked_function_entry, hook_info[2])
-  end
+    local obj, sucess = get_object_reference(obj) --luacheck: ignore
+    if obj and not sucess then
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: *success

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 3]
```diff
@@ -1,367 +1,403 @@
...
-
-VMFMod.hook_remove = function (self, hooked_function_name)
-
-  if not check_function_name(self, "hook_remove", hooked_function_name) then
-    return
-  end
-
-  modify_hook(self, hooked_function_name, "remove")
+-- :hook() will allow you to hook a function, allowing your handler to replace the function in the stack,
+--         and control it's execution. All hooks on the same function will be part of a chain, with the
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: its execution*

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 4]
```diff
@@ -1,367 +1,403 @@
...
 
-  update_function_hook_chain(hooked_function_name)
+    local orig = get_orig_function(self, obj, method)
 
-  _DELAYED_HOOKS[hooked_function_name] = nil
+    if _registry[self][orig] then
+        _registry[self][orig].active = enabled_state
+    else
+        self:warning("(%s): trying to toggle hook that doesn't exist: %s", func_name, method)
+        return
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: Unnecessary return?

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 5]
```diff
@@ -1,367 +1,403 @@
...
+    for _, hook in pairs(_registry[self]) do
+        hook.active = true
+    end
 end
 
-
-VMFMod.enable_all_hooks = function (self)
-  modify_all_hooks(self, "enable")
+function VMFMod:disable_all_hooks()
+    -- Using pairs because the self table may contain nils, and order isnt important.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: isn't*

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 6]
```diff
@@ -1,367 +1,403 @@
...
-
-VMFMod.disable_all_hooks = function (self)
-  modify_all_hooks(self, "disable")
+    
+-- Enable/disable functions for all hook types:
+function VMFMod:hook_enable(obj, method)  generic_hook_toggle(self, obj, method, true) end
+function VMFMod:hook_disable(obj, method) generic_hook_toggle(self, obj, method, false) end
+
+function VMFMod:enable_all_hooks()
+    -- Using pairs because the self table may contain nils, and order isnt important.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: isn't*

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 7]
```diff
@@ -1,367 +1,403 @@
...
+            self:error("(%s): trying to create hook without giving a method name. %s", func_name)
+            return
+        end
+    end
 
-    for _, hook_info in ipairs(_DELAYED_HOOKS[hooked_function_name]) do
+    -- Get hook_type based on name
+    local hook_type = HOOK_TYPES[func_name]
+
+    -- Grab the object's reference, if this fails, obj will remains a string and the hook will be delayed.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: remain*

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 8]
```diff
@@ -1,367 +1,403 @@
...
-        hook_entry.exec_function = hooked_function_entry.hooks[i - 1].exec_function
-      end
+    elseif hook_type == HOOK_TYPE_SAFE then
+        func = function(...)
+            if hook_data.active then
+                vmf.xpcall_no_return_values(self, "safe_hook", handler, ...)
+            end
+        end
+    else
+        self:error("(create_specialized_hook): Invalid hook_type given. You should never this see.")
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: never see this?

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 9]
```diff
@@ -1,367 +1,403 @@
...
+            end
         end
-      else
-        hook_entry.exec_function = hooked_function_entry.original_function
-      end
-    else
-      if hook_entry.is_enabled then
-        hook_entry.exec_function = function(...)
-          return hook_entry.hook_function(hooked_function_entry.hooks[i - 1].exec_function, ...)
+    -- Rawhooks need to directly call the original function is inactive.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: Raw hooks are origin hooks now, right?

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 10]
```diff
@@ -1,367 +1,403 @@
...
-    -- because next function entry will be skiped by 'for'
-    -- so it have to be done later
-    if #hooked_function_entry.hooks == 0 then
-      table.insert(no_hooks_functions_indexes, 1, i)
+    -- Check to make sure it wasn't hooked before
+    if not _registry[self][orig] then
+        _registry[self][orig] = { active = true }
+
+        local hook_registry = _registry.hooks[hook_type]
+        -- Add to the hook to registry. Raw hooks are unique, so we check for that too.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: Raw hooks are origin hooks now, right?

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 11]
```diff
@@ -1,367 +1,403 @@
...
-      if hook_entry.is_enabled then
-        hook_entry.exec_function = function(...)
-          return hook_entry.hook_function(hooked_function_entry.original_function, ...)
+-- Returns a function closure with all the information needed for a given hook to be handled correctly.
+local function create_specialized_hook(self, orig, handler, hook_type)
+    local func
+    local hook_data = _registry[self][orig]
+
+    -- Determine the previous function in the hook stack
+    -- Note: If a previous hook is removed from the table, these functions wouldnt be updated
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: wouldn't*

File: vmf/scripts/mods/vmf/modules/core/hooks.lua [hunk 12]
```diff
@@ -1,367 +1,403 @@
...
-local function get_hook_entry(mod, hooked_function_entry)
+-- ####################################################################################################################
+-- ##### Hook Creation ################################################################################################
+-- ####################################################################################################################
 
-  for i, hook_entry in ipairs(hooked_function_entry.hooks) do
-    if hook_entry.mod == mod then
-      return hook_entry, i
+-- For any given original function, return the newest entry of the hook_chain.
+-- Since all hooks of the chain contains the call to the previous one, we don't need to do any manual loops.
```
Comment (Aussiemon) [inline vmf/scripts/mods/vmf/modules/core/hooks.lua]: contain*

**Review by Aussiemon — CHANGES_REQUESTED** [review]
Very serious merge-blocking issues that aren't nit-picky at all.

**Review by Aussiemon — APPROVED** [review]
